// npm-build-config/dist/npm-build-config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _fs = _interopRequireDefault(require("fs"));

var _path = _interopRequireDefault(require("path"));

var _npmBuildUtil = _interopRequireDefault(require("npm-build-util"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// 第三方
// 组件库
var pkg = require(_path.default.resolve('package.json'));
/************************* 环境变量 ***************************/


var developer = _fs.default.existsSync('./developer.config.js') ? require('../../developer.config.js').default : null;
var env = {
  env: process.env.env !== '[object Object]' && process.env.env || '',
  num: process.env.num || '',
  mini: process.env.env === 'prod' || process.env.mini == 'true',
  check: process.env.check == 'true',
  repair: process.env.repair == 'true'
};
/************************* 字符串 ***************************/

var str = {
  min: env.mini ? '.min' : ''
};
/************************* 文件glob路径 ***************************/

var path = {
  project: _npmBuildUtil.default.getProjectPath(),
  html: ["node_modules/dvd-base-build-node-ssr/html/template.html"],
  css: ["src/*page/".concat(env.page, "/css/*.scss"), "src/*common/css/common.scss"],
  js: "src/page/".concat(env.page, "/js/*.js"),
  router: "src/*page/".concat(env.page, "/js/router.js"),
  // dll: `node_modules/dvd-base-build-node-ssr/dist/dll/*.dll${env.env && '.min'}.js`,
  dll: "node_modules/dvd-base-build-node-ssr/dist/dll/*.js",
  moveJs: "src/*common/js/autoRootSize.js",
  img: ["src/*page/".concat(env.page, "/img/*"), "src/*common/img/*"],
  iconDir: "src/page/".concat(env.page, "/img/icon*"),
  temp: ".temp",
  dist: "dist",
  static: "dist/static",
  include: "".concat(__dirname, "/../../")
};
/************************* 环境参数替换表 ***************************/

var replacer = {
  '[[env]]': env.env,
  '[[num]]': env.num,
  '[[base_domain]]': function () {
    if (env.env == 'dev') {
      return 'bravetime.net';
    } else if (env.env == 'beta') {
      return 'vyohui.cn';
    } else {
      return 'davdian.com';
    }
  }(),
  '[[static]]': function () {
    // 只有生产环境和stage环境有CDN加速
    if (env.env === 'prod') {
      return "https://cdn.yunshuxie.com/".concat(pkg.name);
    }
    /*else if (env.env === 'stage' || env.env === 'gray' || env.env === 'dev') {
    return `https://${pkg.name}.yunshuxie.com/static`;
    }*/
    else {
        return "/static";
      }
  }(),
  '[[vendor]]': '//cdn.yunshuxie.com/vendor',
  '[[v]]': env.mini ? "" : "?v=".concat(_npmBuildUtil.default.getNow()),
  // '[[project]]': pkg.name,
  regex: /\[\[(env|num|base_domain|static|vendor|v|project)]]/g
}; // replacer.regex = function () {
//   let arr = [];
//   for (let i in replacer) {
//     let key = i.replace(/\[/g, '\\[');
//     arr.push(key);
//   }
//   console.log(arr);
//   return new RegExp(arr.join('|'), 'g');
// }();

/************************* css前缀 ***************************/

var autoprefixer = {
  browsers: ["last 4 versions", "Android >= 4", "iOS >= 4", "IE >= 8", "> 0.1%", "Firefox >= 20"]
};
/************************* 获取别名 ***************************/

var getAlias = function getAlias(target) {
  var min = config.env.mini === true ? '.min' : ''; // 为模块添加别名，根据环境动态引用不同js（压缩|非压缩|服务端|客户端）

  var alias = {
    vue$: "vue/dist/vue".concat(min, ".js"),
    swiper$: "swiper/dist/js/swiper".concat(min, ".js")
  }; // 为每个npm-模块添加别名，根据环境动态引用不同js

  _fs.default.readdirSync(_npmBuildUtil.default.findInInnerOrBrother('')).forEach(function (dir) {
    if (dir.startsWith('npm-')) {
      alias["".concat(dir, "$")] = "".concat(dir, "/dist/").concat(dir).concat(target == 'client' ? '.client' : '').concat(min, ".js");
    }
  });

  return alias;
};

var config = {
  env: env,
  str: str,
  path: path,
  replacer: replacer,
  pkg: pkg,
  autoprefixer: autoprefixer,
  getAlias: getAlias
};
console.log("/************************* config\u53C2\u6570\u4E3A ***************************/");
console.log("config: ".concat(_npmBuildUtil.default.stringifyFormat(config)));
var _default = config;
exports.default = _default;
