// npm-build-config/src/npm-build-config.js
// 第三方
import fs from 'fs';
import sysPath from 'path';

// 组件库
import util from 'npm-build-util';

let pkg = require(sysPath.resolve('package.json'));

/************************* 环境变量 ***************************/
let developer = fs.existsSync('./developer.config.js') ? require('../../developer.config.js').default : null;
var env = {
  env: process.env.env !== '[object Object]' && process.env.env || '',
  num: process.env.num || '',
  mini: process.env.env === 'prod' || process.env.mini == 'true',
  check: process.env.check == 'true',
  repair: process.env.repair == 'true',
};

/************************* 字符串 ***************************/
var str = {
  min: env.mini ? '.min' : '',
};

/************************* 文件glob路径 ***************************/
let path = {
  project: util.getProjectPath(),
  html: [`node_modules/dvd-base-build-node-ssr/html/template.html`],
  css: [`src/*page/${env.page}/css/*.scss`, `src/*common/css/common.scss`],
  js: `src/page/${env.page}/js/*.js`,
  router: `src/*page/${env.page}/js/router.js`,
  // dll: `node_modules/dvd-base-build-node-ssr/dist/dll/*.dll${env.env && '.min'}.js`,
  dll: `node_modules/dvd-base-build-node-ssr/dist/dll/*.js`,
  moveJs: `src/*common/js/autoRootSize.js`,
  img: [`src/*page/${env.page}/img/*`, `src/*common/img/*`],
  iconDir: `src/page/${env.page}/img/icon*`,
  temp: `.temp`,
  dist: `dist`,
  static: `dist/static`,
  include: `${__dirname}/../../`,
};

/************************* 环境参数替换表 ***************************/
let replacer = {
  '[[env]]': env.env,
  '[[num]]': env.num,
  '[[base_domain]]': (function () {
    if (env.env == 'dev') {
      return 'bravetime.net';
    } else if (env.env == 'beta') {
      return 'vyohui.cn';
    } else {
      return 'davdian.com';
    }
  })(),
  '[[static]]': (function () {
    // 只有生产环境和stage环境有CDN加速
    if (env.env === 'prod') {
      return `https://cdn.yunshuxie.com/${pkg.name}`;
    } /*else if (env.env === 'stage' || env.env === 'gray' || env.env === 'dev') {
     return `https://${pkg.name}.yunshuxie.com/static`;
     }*/ else {
      return `/static`;
    }
  })(),
  '[[vendor]]': '//cdn.yunshuxie.com/vendor',
  '[[v]]': env.mini ? `` : `?v=${util.getNow()}`,
  // '[[project]]': pkg.name,
  regex: /\[\[(env|num|base_domain|static|vendor|v|project)]]/g,
};
// replacer.regex = function () {
//   let arr = [];
//   for (let i in replacer) {
//     let key = i.replace(/\[/g, '\\[');
//     arr.push(key);
//   }
//   console.log(arr);
//   return new RegExp(arr.join('|'), 'g');
// }();

/************************* css前缀 ***************************/
var autoprefixer = {
  browsers: [
    "last 4 versions",
    "Android >= 4",
    "iOS >= 4",
    "IE >= 8",
    "> 0.1%",
    "Firefox >= 20"
  ]
};

/************************* 获取别名 ***************************/
let getAlias = target => {
  let min = config.env.mini === true ? '.min' : '';

  // 为模块添加别名，根据环境动态引用不同js（压缩|非压缩|服务端|客户端）
  let alias = {
    vue$: `vue/dist/vue${min}.js`,
    swiper$: `swiper/dist/js/swiper${min}.js`,
  };

  // 为每个npm-模块添加别名，根据环境动态引用不同js
  fs.readdirSync(util.findInInnerOrBrother('')).forEach(dir => {
    if (dir.startsWith('npm-')) {
      alias[`${dir}$`] = `${dir}/dist/${dir}${target == 'client' ? '.client' : ''}${min}.js`
    }
  });

  return alias;
};

let config = {
  env,
  str,
  path,
  replacer,
  pkg,
  autoprefixer,
  getAlias,
};

console.log(`/************************* config参数为 ***************************/`);
console.log(`config: ${util.stringifyFormat(config)}`);

export default config;
